// angular-dual-listbox/esm5/lib/basic-list.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var BasicList = /** @class */ (function () {
    function BasicList(name) {
        this._name = name;
        this.last = null;
        this.picker = '';
        this.dragStart = false;
        this.dragOver = false;
        // Arrays will contain objects of { _id, _name }.
        this.pick = [];
        this.list = [];
        this.sift = [];
    }
    Object.defineProperty(BasicList.prototype, "name", {
        get: /**
         * @return {?}
         */
        function () {
            return this._name;
        },
        enumerable: true,
        configurable: true
    });
    return BasicList;
}());
export { BasicList };
if (false) {
    /**
     * Name of the list
     * @type {?}
     * @private
     */
    BasicList.prototype._name;
    /**
     * Last element touched
     * @type {?}
     */
    BasicList.prototype.last;
    /**
     * text filter
     * @type {?}
     */
    BasicList.prototype.picker;
    /** @type {?} */
    BasicList.prototype.dragStart;
    /** @type {?} */
    BasicList.prototype.dragOver;
    /** @type {?} */
    BasicList.prototype.pick;
    /** @type {?} */
    BasicList.prototype.list;
    /** @type {?} */
    BasicList.prototype.sift;
}
