// angular-dual-listbox/esm2015/lib/basic-list.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
export class BasicList {
    /**
     * @param {?} name
     */
    constructor(name) {
        this._name = name;
        this.last = null;
        this.picker = '';
        this.dragStart = false;
        this.dragOver = false;
        // Arrays will contain objects of { _id, _name }.
        this.pick = [];
        this.list = [];
        this.sift = [];
    }
    /**
     * @return {?}
     */
    get name() {
        return this._name;
    }
}
if (false) {
    /**
     * Name of the list
     * @type {?}
     * @private
     */
    BasicList.prototype._name;
    /**
     * Last element touched
     * @type {?}
     */
    BasicList.prototype.last;
    /**
     * text filter
     * @type {?}
     */
    BasicList.prototype.picker;
    /** @type {?} */
    BasicList.prototype.dragStart;
    /** @type {?} */
    BasicList.prototype.dragOver;
    /** @type {?} */
    BasicList.prototype.pick;
    /** @type {?} */
    BasicList.prototype.list;
    /** @type {?} */
    BasicList.prototype.sift;
}
